### DISCUSSION TITLE
Target-typed casting

### DISCUSSION THREAD

**Discussion by vc-74** [dotnet/csharplang#7017] [Language Ideas]
[target-typed new expressions](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/target-typed-new) are a great way to improve code readability. A similar idea could be applied when declaring a variable and assigning it from the result of an explicit cast.

For instance, instead of:
`var list = (IReadOnlyList<int>)collection;`
or 
`IReadOnlyList<int> list = (IReadOnlyList<int>)collection;`

we could write
`IReadOnlyList<int> list = collection!;`

The bang operator is just used to provide a syntax example and a different syntax could be used as long as
- The type to cast to precedes the variable name
- The type to cast to is not repeated in the expression

ed: adding more details following the comments below:

To me target-typed new expressions have been created to improve readability of variable declaration and assignment on the same line and avoid having to duplicate the type.

Before them, we could write:
`Class c = new Class()`
and
`var c = new Class()`

In the first case, it's obvious that `c` is a `Class` since the type precedes the name of the variable, but the type is duplicated.
In the second it is not duplicated but it's a little bit less obvious to know what `c` is.

Target-typed new expressions allow the best of both worlds:
`Class c = new()`

My suggestion is to add a similar feature to other declaration + assignment cases, in particular when declaring a variable and assigning it to the result of an explicit cast.

**Comment by FaustVX** [comment]
Everything you asked already works
```C#
IReadOnlyList<int> list1 = collection;
var list2 = (IReadOnlyList<int>)collection;
IReadOnlyList<int> list3 = (IReadOnlyList<int>)collection;
```
[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAAQCYBGAWACgCAGAAgJIBYBuSy5CdWgYwHsAbfqm4IovGLQC8tGKgDutADJQAzggA8sBAD4AFAEpWFSgEkASqggATAPIx+OZWs0wdtfqoQkpPAUJFiMEYcXB5qRD665pa29o6eLjr6fILCouJG0dZ2Dk4aWtrungDMkVmxuQkFyX5pgcxAA===)

**Comment by vc-74** [comment]
@FaustVX what I'm suggesting here is a new syntax for those casting cases (assuming `collection` is an `IEnumerable<int>` for instance).

**Reply by FaustVX** [reply]
For me, the `list2` works well and is clear enough
`IEnumerable<T>` doesn't implements `IReadOnlyList<T>`, so hiding the conversion can be dangerous. with explicit casting, _you_ tell the compiler that _you_ know what you are doing.

**Reply by vc-74** [reply]
I know, I don't want to make the cast implicit, just to make the syntax shorter while having the type to cast to to the left.

**Reply by HaloFour** [reply]
> I don't want to make the cast implicit

If the cast is not safe enough that the compiler can guarantee that it will be successful then it should definitely not be implicit.  That is the point of explicit casting, making the developer declare their intent that they are willing to live with the consequences of the cast failing at runtime.

**Reply by FaustVX** [reply]
> I know, I don't want to make the cast implicit

If you don't specify the type to cast, then it's an implicit cast.

**Reply by vc-74** [reply]
> If you don't specify the type to cast, then it's an implicit cast.

The type is specified:
`IReadOnlyList<int> list = collection!;`

but as the target type, not using the current C# syntax with parentheses.

**Reply by HaloFour** [reply]
> but as the target type, not using the current C# syntax with parentheses.

Correct, that is just a normal assignment, not a cast.  `var` is only about variable inference, not about type safety.

**Reply by vc-74** [reply]
> If the cast is not safe enough that the compiler can guarantee that it will be successful then it should definitely not be implicit. That is the point of explicit casting, making the developer declare their intent that they are willing to live with the consequences of the cast failing at runtime.

I could not agree more. Again, I'm not requesting a new feature, but suggesting a different syntax for a feature that already exists.

**Reply by HaloFour** [reply]
That is already legal syntax, the `!` post-fix operator is the [null forgiving](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving) operator.

**Reply by jnm2** [reply]
@HaloFour It's okay that that's already legal syntax, so long as this proposal would only change behavior of currently non-compiling code.

**Reply by HaloFour** [reply]
Perhaps, but the only reason that code wouldn't compile before would be because the lack of an implicit cast.  So that eould overload the null forgiveness operator, which very explicitly does not change runtime behavior, with a new operator that does and would be very easy to miss.

IMO explicit casts should be that, explicit and obvious, as the developer declares their intent to the compiler that they know that the cast is narrowing and could fail at runtime.  I don't think the null forgiveness operator would be the right syntax to consider as an alternative syntax for explicit casting.

**Reply by vc-74** [reply]
@HaloFour Do you mean you think that using the suggested syntax makes the cast less obvious?

**Reply by HaloFour** [reply]
@vc-74 

Yes, specifically because null forgiving operators can be used in assignment statements now and that operator is explicit documented to not change runtime behavior.

**Reply by vc-74** [reply]
I used `!` as an example, a different syntax could be used if needed

**Comment by jnm2** [comment]
This would be similar to how `foreach (Bar bar in foos)` casts `foos` items to `Bar`, and same with the LINQ syntax `from Bar bar in foos`.

**Comment by TahirAhmadov** [comment]
The gist of the request, the way I understand it, is that you want `!` operator to have a separate meaning: in addition to the long time boolean negation, and the more recent null forgiveness, you want it to now also mean explicit casting.

The problem is that in those types of expressions the `!` already has a different meaning. With boolean negation, it's used in a different context, which is why it was possible to re-use it for the null forgiveness - it appears at the end of a reference type expressions, as opposed to in front of a boolean typed expression. However, with explicit casting, `!` after a reference type already has a meaning; your new "explicit cast" meaning would directly conflict with existing semantics, making this a non-starter.

Also, even if we were to find a different symbol other than `!` for explicit casting, what's wrong with `var list = (IReadOnlyList<int>)collection;`?

**Reply by vc-74** [reply]
Nothing per se, it's just that with target-typed new expressions the types are all to the left side of the assignment expression.

**Reply by HaloFour** [reply]
Yes, but with target-typed new expressions there is no coercion or cast, the type on the left is the same as the type on the right.  That's not true in this case.

**Comment by jnm2** [comment]
> the syntax of this feature could differ as long as
> 
> * The type to cast to is to the left of the expression
> * The type to cast to is not repeated in the expression

These criteria are already satisfied by the `var` form. Why is this not a great existing solution?

The only thing I could think of was that `!` is postfix and that is convenient. An idea that's been brought up a few times by an LDM member or two is postfix casts (request at https://github.com/dotnet/csharplang/discussions/5447), looking something like:

```cs
var list = collection.(IReadOnlyList<int>);
```

**Reply by vc-74** [reply]
> These criteria are already satisfied by the `var` form. Why is this not a great existing solution?

Not quite, by to the left I mean something similar to target-typed new expressions:

`List<int> list = new()`

**Reply by jnm2** [reply]
Oh, do you mean target-typing rather than relative lexical position? For instance, do you mean this would also work?

```cs
M(collection!);

void M(List<int> list) { }
```

**Reply by vc-74** [reply]
No I don't think it's worth going this far.
To me target-typed new expressions have been created to improve readability of variable declaration and assignment on the same line and avoid having to duplicate the type.

Before them, we could write:
`Class c = new Class()`
and
`var c = new Class()`

In the first case, it's obvious that `c` is a `Class` since the type precedes the name of the variable, but the type is duplicated.
In the second it is not duplicated but it's a little bit less obvious to know what `c` is.

Target-typed new expressions allow the best of both worlds:
`Class c = new()`

My suggestion is to add a similar feature to other declaration + assignment cases, in particular when declaring a variable and assigning it to the result of an explicit cast.

**Comment by Richiban** [comment]
Why do you want to be able to write `IReadOnlyList<int> list = collection!;` when `IReadOnlyList<int> list = collection;` already works (without the bang `!`). Perhaps I've misunderstood your request, but if you're going to write the type on the left then you can assign what you need without casting at all.

**Reply by jnm2** [reply]
It does not work if `collection` is a less-specific type.

**Reply by vc-74** [reply]
> It does not work if `collection` is a less-specific type.

Exactly

**Reply by Richiban** [reply]
Now I'm with you! You want to automatically _downcast_ (or explicit cast?).

I'm pretty sure I've seen a request like this before; IIRC that proposed to allow casting to `var`. So:

```
IEnumerable<string> strings = new List<string>();

IReadOnlyCollection<string> string2 = (var)strings;
```

I'll see if I can find it.

**Reply by Richiban** [reply]
Here: https://github.com/dotnet/csharplang/discussions/4428

**Reply by vc-74** [reply]
Indeed, it looks like a similar feature, thanks. I'm not sure `var` is a good choice of keyword however, it doesn't clearly convey the idea of explicit casting.
